### DISCUSSION TITLE
News: Raspberry Pi Pico (RP2040) support

### DISCUSSION THREAD

**Discussion by Bodmer** [Bodmer/TFT_eSPI#1111] [Raspberry Pi Pico]
I have updated the TFT_eSPI master library (only[url=https://github.com/Bodmer/TFT_eSPI] available from Github[/url] at the moment).

The library has been tested with the Raspberry Pi Pico connected to an ILI9341 4 wire SPI display.

The examples are working and performance is quite good.

For those familiar with the "Graphicstest" example these are the results:
```
Benchmark,                Time (microseconds)
Screen fill (5 times),    114981  (43.49 fps)
Text,                     31986
Lines,                    128851
Horiz/Vert Lines,         18031
Rectangles (outline),     10210
Rectangles (filled),      241497
Circles (filled),         70702
Circles (outline),        58318
Triangles (outline),      33224
Triangles (filled),       109793
Rounded rects (outline),  30585
Rounded rects (filled),   255167
Total = 1103345us
Total = 1.1033s
```

The Jpeg decode is slower that processors that have a dedicated FPU but it is still acceptable.

Sprites work very well and the large RAM means full frame buffers are possible for the smaller displays (e.g. 240x320).

**Comment by Bodmer** [comment]
The official Arduino board package for the Raspberry Pi Pico was released recently. Quite a few warnings get reported but I have updated the library so it compiles and runs fine.

So the library master copy is now compatible with both the official board package and Earle Philhower's.

Update: PIO capability has been edded. The PIO functions only work with Earle Philhower's board package:
https://github.com/earlephilhower/arduino-pico

**Comment by rcla** [comment]
The "Read_User_Setup" sketch should be updated to correctly display the pins used in the Raspberry Pi Pico.

Inside the "getPinName" function, I added:
```c
if (user.esp == 0x2040) return pin; 
```
In the part of displaying the SPI pins:
```c
if (user.esp == 0x2040){
  if (user.pin_tft_miso != -1) { Serial.print("MISO    = "); Serial.print("GPIO "); Serial.println(getPinName(user.pin_tft_miso)); }
  if (user.pin_tft_mosi != -1) { Serial.print("MOSI    = "); Serial.print("GPIO "); Serial.println(getPinName(user.pin_tft_mosi)); }
  if (user.pin_tft_clk  != -1) { Serial.print("SCK     = "); Serial.print("GPIO "); Serial.println(getPinName(user.pin_tft_clk)); }
else
 {
  if (user.pin_tft_mosi != -1) { Serial.print("MOSI    = "); Serial.print("GPIO "); Serial.print(getPinName(user.pin_tft_mosi)); Serial.println(user.pin_tft_mosi); }
  if (user.pin_tft_miso != -1) { Serial.print("MISO    = "); Serial.print("GPIO "); Serial.print(getPinName(user.pin_tft_miso)); Serial.println(user.pin_tft_miso); }
  if (user.pin_tft_clk  != -1) { Serial.print("SCK     = "); Serial.print("GPIO "); Serial.print(getPinName(user.pin_tft_clk)); Serial.println(user.pin_tft_clk); }  
 }
```

**Comment by Dresch123** [comment]
Hi,
Thanks very much for the updates for the Pico Pi. I have most of the examples working including the BMP and TFT display, including reading off an SD card.
For some reason, I cannot get the display of RAW image files working (reading RAW files off an SD). The images come out as multiple small images with weird color distortion. I am using the drawRAW example previously published and RAW files I used successfully with the Adafruit library. 
Could you take a look at pushColors and see if anything has changed for the PicoPi build? Sorry for bothering you but I have been diddling with this for days and can't seem to make it work. Eventually I would like to get RAW video running also.
Thanks,
Bill

**Comment by Dresch123** [comment]
I am attaching three images: one generated with drawBMP, the other one displaying a JPEG and then drawRAW using pushColors.
![dreawBMP](https://user-images.githubusercontent.com/3015603/126090822-9e2fea03-04df-4e52-b0f9-7cc1152d2731.jpg)
![dreawJPG](https://user-images.githubusercontent.com/3015603/126090840-da21c8a8-53da-4a61-8563-d8e882755358.jpg)
![drawRAW](https://user-images.githubusercontent.com/3015603/126090848-41cda100-c096-4627-8b42-a3893d8f5c15.jpg)

**Comment by Dresch123** [comment]
rawFile.read((uint8_t *)&sdbuffer, rawFile.size());
 tft.setSwapBytes(false);
 tft.pushImage(0,0,320,240,sdbuffer);

Works for RAW file display!

**Comment by djairjr** [comment]


**Reply by jeancode** [reply]
es una locura te voy a denunciar con la cia

**Reply by Dresch123** [reply]
> I confirm that the DMA is also working with the GC9A01 display, as seen in the image. However, I haven't been able to make the Animated_Eyes work yet.
> 
> ![RaspiPico GC9A01](https://user-images.githubusercontent.com/1681889/127675136-16438764-7f17-477a-8f79-bada040fe971.jpg)
> 
>  RaspiPico.GC9A01.mp4

Could you post the User_Setups file that you used? I have been trying to get my GC9A01 display working with the Pico Pi... but no luck.
Thanks, Bill

**Reply by djairjr** [reply]
`
//                            USER DEFINED SETTINGS
//   Set driver type, fonts to be loaded, pins used and SPI control method etc
//
//   See the User_Setup_Select.h file if you wish to be able to define multiple
//   setups and then easily select which setup file is used by the compiler.
//
//   If this file is edited correctly then all the library example sketches should
//   run without the need to make any more changes for a particular hardware setup!
//   Note that some sketches are designed for a particular TFT pixel width/height

#define GC9A01_DRIVER // Round TFT SPI Display 

#define TFT_WIDTH  240 // ST7789 240 x 240 and 240 x 320
#define TFT_HEIGHT 240 // GC9A01 240 x 240

#define TFT_MISO 0 // RX PIN - Not Connected
#define TFT_SCLK 2
#define TFT_MOSI 3 // TX PIN
#define TFT_CS   20  // Chip select control pin
#define TFT_DC   18  // Data Command control pin
#define TFT_RST  19  // Reset pin (could connect to Arduino RESET pin)

#define TFT_BL   17            // LED back-light control pin
#define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT
#define PIXEL_DOUBLE // Testando

#define SPI_FREQUENCY  27000000
//#define SPI_FREQUENCY  40000000

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  20000000

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  5000000
// #define SPI_TOUCH_FREQUENCY  2500000 - Copy from Setup200_GC9A01
`

**Reply by Dresch123** [reply]
Thanks very much! My definition looked really similar (same pins even)... I am beginning to wonder if I am having some problem with my Pico Pi hardware.
Did you use Mbed or the Philhower core?
Thanks again! Bill

**Reply by Dresch123** [reply]
Thanks again! Using the MBed core I got the GC9A01 working.  
It is a beautiful display and very low power.

**Comment by Dresch123** [comment]
Has anyone gotten Pico Pi with an ILI9225 display working with TFT_eSPI? 
I know my hardware setup is correct because when I use the library TFT_22_ILI9225: https://github.com/Nkawu/TFT_22_ILI9225
All of those examples run perfectly. 
But the same hardware set up and user definition pins, I get random horizontal lines from TFT_eSPI.
Any suggestions would be appreciated. Thanks,

**Reply by Bodmer** [reply]
The library has been updated and the RP2040 will drive an ILI9225 now.

**Reply by Dresch123** [reply]
Many thanks! I really appreciate you looking into that.

**Reply by Dresch123** [reply]
Sorry, this has taken me a while. I downloaded the newest version of TFT_eSPI and tried it with this user setup:

#define ILI9225_DRIVER

#define TFT_WIDTH  176
#define TFT_HEIGHT 220

//  #define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//  #define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

// For the Pico use these #define lines
#define TFT_MISO  -1
#define TFT_MOSI  19
#define TFT_SCLK  18
#define TFT_CS    17  // Chip select control pin
#define TFT_DC    20  // Data Command control pin
#define TFT_RST   2  // Reset pin (could connect to Arduino RESET pin)


#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT

/ #define SPI_FREQUENCY  20000000
#define SPI_FREQUENCY  10000000
// #define SPI_FREQUENCY  40000000

#define SPI_TOUCH_FREQUENCY  2500000
// #define SUPPORT_TRANSACTIONS

Sadly I get the exact same thing as I described before when I run the TFT_Rainbow example. I tested the same setup again with the TFT_22_ILI9225 driver and those examples run without an issue. Is there anything else I should be doing? Thanks, Bill

**Comment by Bodmer** [comment]
Run the Read_User_Setup sketch and check the serial monitor output. This will indicate if your setup is being used. If not, the check the correct setup is being loaded in User_Setup_Select.h

https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Test%20and%20diagnostics/Read_User_Setup

**Reply by Dresch123** [reply]
This is what I get:
[code]
TFT_eSPI ver = 2.4.43
Processor    = RP2040
Transactions = Yes
Interface    = SPI
Display driver = 9225
Display width  = 176
Display height = 220

MOSI    = GPIO 19
SCK     = GPIO 18
TFT_CS   = GPIO 17
TFT_DC   = GPIO 20
TFT_RST  = GPIO 2

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 10.00
[/code]
I have attached a photo of the working TFT_22_ILI9225 and the non-working "Rainbow" TFT_eSPI example. 
Thanks for looking at this. 
Here is the relevant set-up from TFT_22_ILI9225:
#elif defined (ARDUINO_ARCH_RP2040) // Raspberry Pi Pico
#define TFT_RST 2   // D2
#define TFT_RS  20   // D1
#define TFT_CLK 18
//#define TFT_SDO 12  // D6 MISO
#define TFT_SDI 19  // D7 MOSI
#define TFT_CS  17  // D8 SS
#define TFT_LED 0   // D4    
This looks the same as User_Setup...
![Rainbow](https://user-images.githubusercontent.com/3015603/161948812-ee38dff5-99e4-4624-9483-78a47184f1d6.jpg)
![IILI9225](https://user-images.githubusercontent.com/3015603/161949317-ff29e620-c268-4823-9127-4daa1265e036.jpg)
.

**Comment by Bodmer** [comment]


**Reply by Bodmer** [reply]
The display also runs the graphicstest reliably with a pixel SPI data rate of 62.5MHz:
```
Benchmark,                Time (microseconds)
Screen fill (5 times),    54864  (91.13 fps)
Text,                     36786
Lines,                    155501
Horiz/Vert Lines,         6733
Rectangles (outline),     5901
Rectangles (filled),      90783
Circles (filled),         63380
Circles (outline),        56750
Triangles (outline),      45111
Triangles (filled),       70405
Rounded rects (outline),  27787
Rounded rects (filled),   107883
Total = 0.7219s

```

**Comment by Bodmer** [comment]
The RP2040 PIO impementation of the SPI interface built into TFT_eSPI does seem to meet the timing constraints imposed by the commands to the ILI9225. This means the existing unmodified library version should work up to 32.25MHz with the ILI9225 if the following is included in the setup file:
```
#define RP2040_PIO_SPI
```
Pixels start getting lost at 62.5MHz with the PIO implementation, but that is pushing the display hard! (graphicstest completes in  a mere 0.39s!).

**Reply by Dresch123** [reply]
Thanks, I will try this tomorrow.
Is there a way of setting the SPI mode (0 thru 3) for the RP2040 with a simple define like this?

**Comment by Dresch123** [comment]
Thanks! I will give that a try tomorrow.
Is there a simple way of setting the SPI mode (0 thru 3) with a define like this?

**Reply by Dresch123** [reply]
Thanks Bodmer. That suggestion of using the RP2040_PIO_SPI define worked great. Using the 32.25MHz SPI I got 47fps with the bouncy ball example.
![ILI9225_Bouncy_47fps](https://user-images.githubusercontent.com/3015603/162547956-e6edcd4f-ded2-4b37-8711-fb0b4ced154f.jpg)
That works great for the Philhower core but the official core does not support the PIO SPI. I tried the Mbed core first and it told me to use the other core for PIO SPI support. I think publishing your other solution would be good also. Thanks for the insight and getting this running so quickly.

**Comment by Dresch123** [comment]
Thanks! I will try this tomorrow.
Is there a way to set the SPI mode (0 thru 3) with a define like this?

**Reply by Bodmer** [reply]
No, SPI mode is fixed at 0 for all displays except ST7789 since I found my display only worked in mode 3. See [here.](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L93-L98)

A simple check could be added to make the mode defined in a user setup but there has not been a need for that yet:
```
#ifndef TFT_SPI_MODE 
  #if defined(ST7789_DRIVER) || defined(ST7789_2_DRIVER)
    #define TFT_SPI_MODE SPI_MODE3
  #else
    #define TFT_SPI_MODE SPI_MODE0
  #endif
#endif
```

**Reply by Dresch123** [reply]
I hope you don't mind me bringing up the ST7789 without CS issue again but I found a better workaround than CS hardware change. And that conversation is locked.
My display worked with SPI Mode 2. I changed the Mode Set like this:
  // Processor specific code used by SPI bus transaction begin/end_tft_write functions
//  #define SET_BUS_WRITE_MODE spi_set_format(SPI_X,  8, (spi_cpol_t)(TFT_SPI_MODE >> 1), (spi_cpha_t)(TFT_SPI_MODE & 0x1), SPI_MSB_FIRST)
//XXXX Temp fix
#define SET_BUS_WRITE_MODE spi_set_format(SPI_X,  8, (spi_cpol_t)1, (spi_cpha_t)0, SPI_MSB_FIRST)  
So with no CS hardware change on this display, I got:
![ST7789_SPI1_MODE2](https://user-images.githubusercontent.com/3015603/162550202-258b5bc9-8467-4e9b-9903-ae77366f1da1.jpg)
Thanks again!

**Reply by Bodmer** [reply]
That's interesting. The data sheet indicates mode 0 is correct (clock idle low, rising edge sample). Mode 3 (clock idle high, falling edge sample) is the only one I found worked on ESP8266 and ESP32.

Mode 0 should be correct, so this all points to startup conditions/state of the clock line being important for the ST7789. Because there is no CS line to sync the correct clock transition to the correct data bit. You would expect a reset to work after setting the SPI mode but it doesn't. It may be the RP2040 board package always switches to mode 0 after the library closes the transaction, causing a clock idle state change and extra clock edge.

**Reply by Dresch123** [reply]
I came across the Mode 2 setting here:
*********************************************************************
https://forums.raspberrypi.com/viewtopic.php?t=300583

[Re: Interfacing PICO with LCD ST7789](https://forums.raspberrypi.com/viewtopic.php?t=300583#p1816094)
[Tue Feb 09, 2021 12:17 am](https://forums.raspberrypi.com/viewtopic.php?p=1816094&sid=79186048f4f581ba70b2ccfbc63100fd#p1816094)
Which st7789 display do you have?

I'm asking because I couldn't get the pio example to work - I had to code up a standard spi hack to get it to run.

I had to set the polarity as

spi_set_format(SPI_PORT,8,SPI_CPOL_1,SPI_CPHA_0,SPI_MSB_FIRST);

before my display would work.

which makes me think some of the st7789's have alternate polarity or something?
**************************************************************************************************
It seems the ifndef decision tree could include ST7789 and RP2040. I think I will implement that way.
I hope to get in more of these small displays in soon to confirm.
Thanks again Bodmer.

**Reply by Bodmer** [reply]
Interesting again, the Raspberry Pi PIO example worked fine on my display that has a CS line. This does suggest display differences, another explanation is that idle state changes are occurring between transactions, hence the variation associated with using different processors.

**Comment by Bodmer** [comment]
New PIO code has been added at v2.3.45 for the RP2040 to allow the fast SPI PIO capability of the RP2040 to be used with 18bit colour SPI displays such as the ILI9488.  Example setup:
```
#define RP2040_PIO_SPI   // Note: TFT_eSPI touch functions not available when PIO used.

#define ILI9488_DRIVER

#define TFT_MISO  0
#define TFT_MOSI  3
#define TFT_SCLK  2
#define TFT_CS   20
#define TFT_DC   22
#define TFT_RST  21

#define LOAD_GLCD
#define LOAD_FONT2
#define LOAD_FONT4
#define LOAD_FONT6
#define LOAD_FONT7
#define LOAD_FONT8
#define LOAD_GFXFF

#define SMOOTH_FONT

#define SPI_FREQUENCY  63000000

```
